### ISSUE TITLE
Attach volume incorrect time report

### ISSUE THREAD

**Issue by jichenjc** [terraform-provider-openstack/terraform-provider-openstack#1038]
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see https://www.terraform.io/community.html.

### Terraform Version
Run `terraform -v` to show the version. If you are not running the latest version of Terraform, please upgrade because your issue may have already been fixed.

master version 
### Affected Resource(s)
Please list the resources as a list, for example:
openstack_compute_volume_attach_v2

If this issue appears to affect multiple resources, it may be an issue with Terraform's core, so please mention this.

### Terraform Configuration Files
```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key.
```

### Debug Output
Please provide a link to a GitHub Gist containing the complete debug output: https://www.terraform.io/docs/internals/debugging.html. 

In addition, running Terraform with the `OS_DEBUG=1` environment variable set will print additional information specific to the OpenStack provider which will be helpful for troubleshooting.

Please do NOT paste the debug output in the issue; just paste a link to the Gist.

### Panic Output
If Terraform produced a panic, please provide a link to a GitHub Gist containing the output of the `crash.log`.

### Expected Behavior
What should have happened?

we are attaching 2 Volumes to VM and the 
openstack_compute_volume_attach_v2.db2: Creating...
  device:      "" => "<computed>"
  instance_id: "" => "9ebed95a-81fa-4a9b-aa70-73829b85015e"
  region:      "" => "<computed>"
  volume_id:   "" => "cb8e2089-550d-4036-9f4e-595076d8d643"
openstack_compute_volume_attach_v2.db: Creating...
  device:      "" => "<computed>"
  instance_id: "" => "9ebed95a-81fa-4a9b-aa70-73829b85015e"
  region:      "" => "<computed>"
  volume_id:   "" => "0a0d4d79-e20a-43e6-a59f-b3d8d2616c22"
openstack_compute_volume_attach_v2.db2: Creation complete after 6s (ID: 9ebed95a-81fa-4a9b-aa70-73829b85015e/cb8e2089-550d-4036-9f4e-595076d8d643)
openstack_compute_volume_attach_v2.db: Still creating... (10s elapsed)
openstack_compute_volume_attach_v2.db: Still creating... (20s elapsed)
openstack_compute_volume_attach_v2.db: Still creating... (30s elapsed)
openstack_compute_volume_attach_v2.db: Still creating... (40s elapsed)
openstack_compute_volume_attach_v2.db: Creation complete after 46s (ID: 9ebed95a-81fa-4a9b-aa70-73829b85015

The first attach request comes at "20:40:22" with  "POST /v2.1/7ae32c32d1f14f68a961e03092fb0d4b/servers/9ebed95a-81fa-4a9b-aa70-73829b85015e/os-volume_attachments ". Than nova receives the query  command for this volume  "GET /v2.1/7ae32c32d1f14f68a961e03092fb0d4b/servers/9ebed95a-81fa-4a9b-aa70-73829b85015e/os-volume_attachments/cb8e2089-550d-4036-9f4e-595076d8d643", and get 200 status code,   6s later "20:40:28", no more request from terraform runner for volume "cb8e2089-550d-4036-9f4e-595076d8d643", just like terraform said, the attach complete after 6s.

The second attach request is at "20:41:01" for request  "POST /v2.1/7ae32c32d1f14f68a961e03092fb0d4b/servers/9ebed95a-81fa-4a9b-aa70-73829b85015e/os-volume_attachments", and 6s later (20:41:07), no more query request from terraform runner for volume id "0a0d4d79-e20a-43e6-a59f-b3d8d2616c22", which means that the attach action complete after 6s.2020-08-03 20:40:22.457 34302 INFO nova.osapi_compute.wsgi.server [req-2eb0fc93-a2b5-4bd3-8d6f-3adc6d5c270a e4b2b89871054df1ad67bb60b212b077e0d2d40cfbbedf79433a1c

### Actual Behavior
What actually happened?

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform apply`

### Important Factoids
Is there anything atypical about your accounts that we should know? For example: Running in EC2 Classic? A custom version of OpenStack? Tight ACLs?

### References
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:
- GH-1234

**Comment by fgimenezm** [comment]
Looking at terraform code it looks like it's expecting the "os-volume_attachments" GET call with "404" while the attachment is ongoing at the backend/hypervisor and with "200" once the attachment is completed.
But this is not how this particular OpenStack instance is responding. It's responding with 200 as soon as the POST is processed and while the attachment is still ongoing in the backend/hypervisor.

I'm not sure what the "os-volume_attachments" expected behaviour is.
I tried to find documentation about what should the "os-volume_attachments" API call respond while the attachment operation is still ongoing on the backend but I couldn't find it.

**Comment by jichenjc** [comment]
@fgimenezm 

the POST send request to API layer and because it's a `cast` (not a `call`) ,it's async action
https://github.com/openstack/nova/blob/master/nova/compute/rpcapi.py#L551
so POST return with 200 (actually it should be `202` I believe) before the attachment is really completed

the condition checked at Terraform is whether the attachment record is created 
https://github.com/openstack/nova/blob/384041daaad27188075008e359be976a9c198a16/nova/compute/api.py#L4533
will create that record, but again, the actually attach not completed yet...

I will paste more data here after further analysis

**Comment by fgimenezm** [comment]
@jichenjc , I think this diagram will clarify the current behaviour that I'm seeing in our tests:

<img width="722" alt="image" src="https://user-images.githubusercontent.com/8161434/89538079-81daf000-d7d0-11ea-9392-4aa589df2dc4.png">

**Comment by fgimenezm** [comment]
This seems related to issue #277

**Comment by jichenjc** [comment]
@fgimenezm I think I might know how to fix it 

we need switch code logic
maybe first check whether the attachment is ready like what we did now

https://github.com/terraform-provider-openstack/terraform-provider-openstack/blob/e33bbfb06108f6b3538ee8d61a97524149ede439/openstack/compute_volume_attach_v2.go#L26

2) add a client for cinder and refer to 
https://docs.openstack.org/api-ref/block-storage/v3/index.html?expanded=list-accessible-volumes-with-details-detail#volumes-volumes

for volume status to be `in-use` ..

**Comment by ozerovandrei** [comment]
It was fixed in #1106 and this fix will be available in `1.34.0` version which I guess will be released this month.
